Update SLEAP module guide for the new pytorch version (#81)
neuroinformatics-unit/HowTo: docs/source/data_analysis/HPC-module-SLEAP.md

Reviewer: ```suggestion
# Use SLEAP on the SWC HPC cluster
```

To learn more about accessing the HPC via SSH, see the [relevant how-to guide](ssh-cluster-target).

### Access to the SLEAP module

> Once you are on the HPC gateway node, you can see the available SLEAP modules by running `module avail SLEAP`:

Reviewer: ```suggestion
Once you are on the HPC gateway node, you can list the available SLEAP modules:
```

```{code-block} console
$ module avail SLEAP
----------------------- /ceph/apps/ubuntu-24/modulefiles-----------------------
   ...  SLEAP/2025-09-30    SLEAP/2026-05-08 (D)

  Where:
   D:  Default Module
...
```

> - `SLEAP/2026-05-08` corresponds to `SLEAP v.1.6.3` ([PyTorch backend](https://docs.sleap.ai/)) — this is the recommended module for all new projects, and what this guide documents.

Reviewer: ```suggestion
- `SLEAP/2026-05-08` corresponds to [SLEAP v1.6.3](https://docs.sleap.ai/v1.6.3/) (PyTorch backend) — this is the recommended module for all new projects, and what this guide documents.
```
Unless you meant to link to a specific page about PyTorch backend?

> - Older modules use the legacy TensorFlow backend (e.g. `SLEAP/2025-09-30` is `SLEAP v.1.3.4`). Load these by full name if you need to maintain compatibility with an existing project, and refer to the [legacy SLEAP documentation](https://legacy.sleap.ai/). Modules with dates before `2025-09-30` are no longer recommended (built for an older Ubuntu).

Reviewer: ````suggestion
- Older modules use the [legacy TensorFlow backend](https://legacy.sleap.ai/). Use these only if you need compatibility with an existing project. 
- Modules dated before `2025-09-30` were built for an older Ubuntu base and are no longer recommended.

To load the recommended PyTorch-based version (the default):

```{code-block} console
$ module load SLEAP
```

To load an older TensorFlow-based version (e.g. `SLEAP/2025-09-30`, which is SLEAP v1.3.4), specify its full name:

```{code-block} console
$ module load SLEAP/2025-09-30
```
````

Reviewer: may also be useful to keep a table mapping module name to version somewhere in this section? although it does mean the table needs to be updated when newer SLEAP modules get added

If a module has been successfully loaded, it will be listed among

> other loaded modules when you run `module list`:

Reviewer: ```suggestion
To view the modules that are currently loaded:
```

```{code-block} console
$ module list

> Currently Loaded Modulefiles:

Reviewer: ```suggestion
Currently Loaded Modules:
```

> ...   15) SLEAP/2026-05-08

Reviewer: ```suggestion
...   1) SLEAP/2026-05-08
```

Author: I will actually leave it as 15 because that's how it shows up for me. There are a bunch of other modules loaded by default before that (things like GCC and SQLite).

The number may change, but I mainly wanted to signal that SLEAP won't necessarily show up as the first on that list.

```

If you have troubles with loading the SLEAP module,

> see this guide's [Troubleshooting section](#problems-with-the-sleap-module).

Reviewer: ```suggestion
see the [](troubleshooting) section.
```

:::{note}
The SLEAP CLI commands `sleap train` and `sleap track` are aliases for
`sleap-nn train` and `sleap-nn track` respectively; the two forms work
interchangeably. For a full list of arguments, run `<command> --help`
(with the SLEAP module loaded) or consult the SLEAP-NN documentation on
[training](https://nn.sleap.ai/latest/guides/training/),
[inference](https://nn.sleap.ai/latest/guides/inference/) and
[tracking](https://nn.sleap.ai/latest/guides/tracking/).

> :::

Reviewer: maybe this fits better at the end of "Model inference" section, just before "The training-inference cycle" where users would have seen both `sleap train` and `sleap track` ?

Author: Agreed. I've moved it there.

### Install SLEAP on your local PC/laptop
While you can delegate the GPU-intensive work to the HPC cluster,
you will need to use the SLEAP GUI for some steps, such as labelling frames.
Thus, you also need to install SLEAP on your local PC/laptop.

We recommend following the official [SLEAP installation guide](https://docs.sleap.ai/latest/installation/).

> To minimise the risk of issues due to incompatibilities between versions, ensure the version of your local installation of SLEAP matches the one you plan to load in the cluster.

Reviewer: ```suggestion
To avoid compatibility issues, make sure your local SLEAP version matches the module you plan to use on the cluster.
The guide also includes a [version compatibility table](https://docs.sleap.ai/latest/installation/#version-compatibility) showing which versions of `sleap`, `sleap-io`, and `sleap-nn` belong together—use this when installing a specific version or updating pinned packages.
```

For, example, to match the latest SLEAP module at the time of writing (`SLEAP/2026-05-08`),

> you will need to run the following command in your local terminal:

Reviewer: ```suggestion
For example, to match the current default module (`SLEAP/2026-05-08`, SLEAP v1.6.3), you can install the corresponding versions locally with:
```

(prepare-the-training-job)=
### Prepare the training job
Follow the [SLEAP tutorial](https://docs.sleap.ai/latest/tutorial/overview/) till
the end of the section on [Initial Labelling](https://docs.sleap.ai/latest/tutorial/initial-labeling/).

> Ensure that the project file (e.g. `labels.v002.slp`) is saved in the mounted SWC filesystem

Reviewer: is there a specific reason for changing to `labels.v002.slp` throughout?

Author: Not really. It's just that I was writing running through the steps myself as I was writing this guide, and the particular file I was using was named like that. Since I was copy-pasting outputs (such as `ls`) from my terminal, I thought it easiest to keep that name throughout.

But I can easily update everything to `labels.v001.slp` now. Though I'm not sure it matters much. In practice people may use any version number here, or any other file name.

Author: Another reason for sticking with `labels.v002.slp` is that it's what they'll see if they go to `/ceph/scratch/neuroinformatics-dropoff/SLEAP_HPC_test_data`. 

I may follow this guide as is during HPC teaching to SWC PhD students, and we may work with that folder directly.

Reviewer: All good, cheers!

```
Navigate to the training job folder (replace with your own path) and list its contents:
```{code-block} console
$ cd /ceph/scratch/neuroinformatics-dropoff/SLEAP_HPC_test_data
$ cd labels.v002.slp.training_job
$ ls -1
centered_instance.yaml
centroid.yaml
inference-script.sh
jobs.yaml
labels.v002.pkg.slp
train-script.sh
```

The YAML configuration files specify the model architecture, training hyperparameters,
and data pipeline settings for each model. You can inspect them with
`cat centroid.yaml` or open them in a text editor.

The precise files will depend on the model configuration you chose in SLEAP.
Here we see two config files, one for the 'centroid' and another for
the 'centered_instance' model. That's because in this example we have chosen

> the 'Top-Down' configuration, which consists of two neural networks - the first

Reviewer: ```suggestion
the ['Top-Down](https://nn.sleap.ai/latest/reference/models/#top-down)' configuration, which consists of two neural networks - the first
```

for isolating the animal instances (by finding their centroids) and the second
for predicting all the body parts per instance.

> ![Top-Down model configuration](https://legacy.sleap.ai/_images/topdown_approach.jpg)

Reviewer: ```suggestion
![Top-Down model configuration](https://nn.sleap.ai/latest/assets/images/topdown_approach.jpg)
```

Although the 'Top-Down' configuration was designed with multiple animals in mind,
it can also be used for single-animal videos. It makes sense to use it for videos

> where the animal occupies a relatively small portion of the frame - see

Reviewer: ```suggestion
it can also be used for single-animal videos,
where the animal occupies a relatively small portion of the frame - see
```

[Model Configuration](https://nn.sleap.ai/latest/reference/models/) for more info.
:::

SLEAP also generates a `train-script.sh` file in the training job folder.
You can inspect it with `cat train-script.sh` to see the training commands it contains —
these are useful as a reference, but they reflect the paths on the machine that
exported the training job package and may not work as-is on the HPC cluster.

> Instead, we'll write the `sleap train` commands from scratch in the next step.

Reviewer: The config dir shows as `.` for me, e.g.:
`sleap train --config-name centroid.yaml --config-dir . trainer_config.ckpt_dir='models' trainer_config.run_name='260629_120424.centroid.n=20'`
I copied both `sleap train` commands from `train-script.sh` for both models into `train_slurm.sh` but got 
> hydra.errors.OverrideParseException: mismatched input '=' expecting <EOF>

After removing the offending `n=20` in `trainer_config.run_name` in both commands, e.g.:
`sleap train --config-name centroid.yaml --config-dir . trainer_config.ckpt_dir='models'`, training works.

Author: Thanks for bringing this up. For me there were 2 problems when I tried to use the `trains-script` as is, the hydra one you also got, and another problem with the paths. That's why I thought it safest to have people write their own train command from scratch.

Let me know if you have another suggestion for handling this.

Author: Re the path problem, I was getting absolute paths in trainer_config.ckpt_dir (e.g. `trainer_config.ckpt_dir='/mnt/Data/sleap-tutorial-data/models'`) which is a path on my laptop.

Not sure why you ended up with `models` (may be an OS-specific difference). I think it's safest to still write the script from sractch,

Author: I've now reworded this paragraph as follows:

> SLEAP also generates a `train-script.sh` file in the training job folder.
You can inspect it with `cat train-script.sh` to see the training commands it contains.
These are useful as a reference, but be cautious about copying them verbatim.
They may point to folders on the machine that exported the training package,
rather than on the cluster. They may also include a `trainer_config.run_name=...`
setting whose value contains an `=` sign. This makes SLEAP stop with an error like
`mismatched input '=' expecting`. Instead, we'll write the `sleap train` commands from
scratch in the next step, letting SLEAP name each training run automatically.

Reviewer: thanks this is good to know and glad we are cross-checking across different platforms!

# Define directories for SLEAP project and exported training job
SLP_DIR=/ceph/scratch/neuroinformatics-dropoff/SLEAP_HPC_test_data

> SLP_JOB_NAME=labels.v002.slp.training_job

Reviewer: i'm not sure we need to define these dirs. If we ask people to put train_slurm.sh in the unzipped training job folder, the slurm script would be minimal:
```sh
# Print GPU info
nvidia-smi

# Load the SLEAP module
module load SLEAP

# Run the training for each model
sleap train --config-name centroid.yaml --config-dir . trainer_config.ckpt_dir='models'
sleap train --config-name centered_instance.yaml --config-dir . trainer_config.ckpt_dir='models'
```

alternatively, we could ask people to edit train_script.sh (no need to create separate train_slurm.sh and chmod +x) by 
- adding the #SBATCH lines, `nvidia-smi`, `module load SLEAP`
- removing `trainer_config.run_name` args in both sleap train commands

Author: That's a great simplification indeed, thanks!

I will go with your first suggestion, as opposes to the alternative (editing SLEAP's script directly may still leave people with path and hydra problems, as mentioned above).

Author: I've implemented your simplified slurm script and adapted the rest of the guide accordingly: [060e361](https://github.com/neuroinformatics-unit/HowTo/pull/81/commits/060e36172d7ab688e2b3c3a256fc6721979aa534)

  (which SLEAP uses internally) requires this form rather than `--ntasks` or `-n`;
  Lightning then manages GPU parallelism within that single process.
  `--cpus-per-task=8` allocates CPU cores to that process for data loading and preprocessing.

- `--gres gpu:a100:1` requests 1 GPU of type A100. To request any available GPU, use `--gres gpu:1`.
  Inspect available GPU types by listing the nodes in the `gpu` and `gpu_lowp` partitions:
    ```{code-block} console
    $ sinfo -p gpu,gpu_lowp -o "%N %G" --noheader
    ```
    Look for the string between `gpu:` and the next `:` (e.g. `a100`, `l40s`).
    Avoid GPUs with CUDA compute capability below 7.5 (unsupported by PyTorch ≥ 2.5);
    at the time of writing, only `p5000` cards are incompatible.
    See the [SWC wiki](https://liveuclac.sharepoint.com/sites/SSC/SitePages/SSC-CPU-and-GPU-Platform-architecture-165449857.aspx)
    and the [NVIDIA CUDA GPUs page](https://developer.nvidia.com/cuda/gpus) for compute capabilities.

- `module load SLEAP` loads the latest SLEAP module and its dependencies.
  PyTorch bundles its own CUDA runtime, so no separate `cuda` module is needed.

> - `cd $SLP_JOB_DIR` is needed because `--config-dir .` in the `sleap train` commands

Reviewer: if you decide to drop defining dirs from the example SLURM script, this can be removed as well

# Load the SLEAP module
module load SLEAP

# Define directory for SLEAP project
SLP_DIR=/ceph/scratch/neuroinformatics-dropoff/SLEAP_HPC_test_data

# Make a directory to store the predictions (if it doesn't exist already)
mkdir -p $SLP_DIR/predictions

# Run the inference command
sleap track \
    -i $SLP_DIR/mice.mp4 \
    -m $SLP_DIR/models/260512_151547.centroid.n=46 \
    -m $SLP_DIR/models/260512_151547.centered_instance.n=46 \
    -d auto \
    -b 4 \
    --tracking \
    -o $SLP_DIR/predictions/labels.v002.predictions.slp
```
The script mirrors the training one, with a few differences:
- `sleap train` is replaced by a single `sleap track` call (split across lines with `\` for readability).
- The time limit `-t` is lower, since inference is typically faster than training (depends on video length and number of models).

> - `--cpus-per-task` and `--mem` are higher; tune these to your specific job, ideally after a scaled-down trial run.

Reviewer: Following order of when these appear in the script

```suggestion
- `--cpus-per-task` and `--mem` are higher; tune these to your specific job, ideally after a scaled-down trial run.
- The time limit `-t` is lower, since inference is typically faster than training (depends on video length and number of models).
- `sleap train` is replaced by a single `sleap track` call (split across lines with `\` for readability).
```

```{code-block} console
$ sbatch infer-slurm.sh
$ squeue --me
```
Upon completion, a `labels.v002.predictions.slp` file will have been created in the `predictions` directory.

You can use the SLEAP GUI on your local machine to load and view the predictions:
*File* -> *Open Project...* -> select the `labels.v002.predictions.slp` file.

## The training-inference cycle

Now that you have some predictions, you can keep improving your models by repeating
the training-inference cycle.

This predictions file has the same format as a standard SLEAP project file,
and you can use the GUI (on your local machine) to manually correct the predictions
or merge them into an existing SLEAP project.

For example, you can:

- [Manually correct](https://docs.sleap.ai/latest/tutorial/correcting-predictions/) some of the predictions
- Merge corrected labels into the initial training set (*File* -> *Merge into Project...*).
- Save the merged training set under a new name, e.g. `labels.v003.slp`
- Export a new training job `labels.v003.slp.training_job` (you may reuse the training configurations from before)
- Repeat the training-inference cycle until satisfied

> ## Troubleshooting

Reviewer: ```suggestion
(troubleshooting)=
## Troubleshooting
```

(legacy-modules)=
## Legacy (TensorFlow) modules

If you are using a legacy SLEAP module (≤ 1.4.1, TensorFlow backend), the
CLI uses `sleap-train` and `sleap-track` with JSON config files instead of
the YAML-based `sleap train` / `sleap track` shown above. See the
[legacy SLEAP documentation](https://legacy.sleap.ai/) for full details;
the equivalents for the steps in this guide are below.

### Training

```{code-block} bash
sleap-train centroid.json labels.v002.pkg.slp
sleap-train centered_instance.json labels.v002.pkg.slp
```

The exported training job package from legacy SLEAP also includes a
`train-script.sh` containing these commands, so you can run
`./train-script.sh` from the SLURM script. See the legacy
[remote training guide](https://legacy.sleap.ai/guides/remote.html#remote-training)
and [CLI reference](https://legacy.sleap.ai/guides/cli.html).

### Inference

```{code-block} bash
sleap-track video.mp4 \
    -m models/centroid/training_config.json \
    -m models/centered_instance/training_config.json \
    --gpu auto \
    --tracking.tracker simple \
    --tracking.similarity centroid \
    -o predictions.slp
```

### Verifying the module

> The verification steps in the [Troubleshooting section](#problems-with-the-sleap-module)

Reviewer: ```suggestion
The verification steps in the [](troubleshooting) section
```
